pkg/kubelet/apis/config/types.go

```diff
@@ -407,6 +423,17 @@ type KubeletConfiguration struct {
 	// class value that lies in the range of that value and the next higher entry in the
 	// list when the node is shutting down.
 	ShutdownGracePeriodByPodPriority []ShutdownGracePeriodByPodPriority
+	// GracefulNodeShutdownPodPolicy defines policy for how pods are handled during graceful shutdown.
+	// The policy types available are `SetTerminal` (default) or `LeaveRunning`.
+	// `SetTerminal` will result in pods being to set terminal phase during shutdown (enabled by default).
+	// If a node becomes ready again after graceful node shutdown (e.g. a reboot), will result in pods not being started on the same node
+	// since the pods will be placed into terminal phase. Pods will be placed in the Failed terminal phase on the API server.
+	// In contrast, with `LeaveRunning` policy, pods will not be placed into terminal phase,
+	// and as result they will resume to run after the node becomes ready again.
```

Comment:
I would probably say this should document that the default behavior for Kubernetes is `LeaveRunning`.  Also, we should describe how the Kubelet detects that the node is being shut down, since the sequence here was:

1. shutdown of node didn't do anything
2. we added a feature that behaved like SetTerminal (different than 1)
3. we realized we needed 1, so we added this flag
4. users seeing this flag don't know how to go from 1 (the way it worked from 1.0 to 1.24) to 3 correctly, or they might be confused that 2 is how it works.

Reply:
> I would probably say this should document that the default behavior for Kubernetes is LeaveRunning

That's a good point. Requiring opt in for the GracefulNodeShutdown feature to mark pods as terminal makes sense to me.

> Also note - I don't expect this flag to have a default value per se - I expect the presence of this flag OR shutdown grace period to turn the feature on (makes us deviate from behavior 1), and thus this value is only defaulted to SetTerminal when its omitted AND ShutdownGracePeriod is set.

That's another good point. I think that means we wouldn't use official API defaulting, but would instead accept "" and treat it like `LeaveRunning` when initializing the kubelet runtime options

Reply:
Thanks @smarterclayton and @liggitt for the feedback.

> Requiring opt in for the GracefulNodeShutdown feature to mark pods as terminal makes sense to me. 

This is already the case, `GracefulNodeShutdownPodPolicy` will only be used if graceful node is enabled (i.e. `ShutdownGracePeriod` is set). Otherwise, if graceful node shutdown is not enabled, shutdown of node won't do anything.

Does it make sense as follow up to then:

1) Only default `GracefulNodeShutdownPodPolicy` to `SetTerminal` if Graceful Node Shutdown feature is enabled (i.e. `ShutdownGracePeriod`) in `pkg/kubelet/apis/config/v1beta1/defaults.go`. Otherwise the setting will be left unset, `""`. 
2) Update the API docs to make it clear that `GracefulNodeShutdownPodPolicy` will be defaulted to `SetTerminal` only if Graceful Node shutdown is enabled.

Reply:
Updated as mentioned above. Please take another look, thanks!

Reply:
> Only default GracefulNodeShutdownPodPolicy to SetTerminal if Graceful Node Shutdown feature is enabled (i.e. ShutdownGracePeriod) in pkg/kubelet/apis/config/v1beta1/defaults.go. Otherwise the setting will be left unset, "".

Let me translate this to make sure we're all on the same page with possible configurations

1. User specifies `ShutdownGracePeriod` but `GracefulNodeShutdownPodPolicy` is empty or null -> defaults to `SetTerminal` (preserves behavior for beta users)
2. User specifies `GracefulNodeShutdownPodPolicy` but `ShutdownGracePeriod` is empty or zero -> default to `LeaveRunning`
3. User specifies `GracefulNodeShutdownPodPolicy` and `ShutdownGracePeriod` -> no default necessary

If we did this, then I think users who move from 2 to 3 would be confused (they'd switch from LeaveRunning to SetTerminal, which is bad).

So I might argue that 2 should be:

2. User specifies `GracefulNodeShutdownPodPolicy` but `ShutdownGracePeriod` is empty or zero -> validation error, user must explicitly specify `ShutdownGracePeriod`

Which then trains new users to make a choice (and we can say `LeaveRunning is the default behavior for pods on kubelet shutdown when grace period is disabled` or something).